File: core/src/main/java/com/batchsphere/core/qms/changecontrol/entity/ChangeControl.java
package com.batchsphere.core.qms.changecontrol.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "qms_change_control")
@Getter @Setter @Builder @NoArgsConstructor @AllArgsConstructor
public class ChangeControl {

    @Id
    @Column(nullable = false, updatable = false)
    private UUID id;

    @Column(name = "change_control_number", nullable = false, unique = true, length = 100)
    private String changeControlNumber;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "change_type", nullable = false, length = 40)
    private ChangeControlType changeType;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String reason;

    @Enumerated(EnumType.STRING)
    @Column(name = "risk_classification", nullable = false, length = 20)
    private ChangeControlRisk riskClassification;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private ChangeControlStatus status;

    @Column(name = "impact_assessment", columnDefinition = "TEXT")
    private String impactAssessment;

    @Column(name = "implementation_plan", columnDefinition = "TEXT")
    private String implementationPlan;

    @Column(name = "effectiveness_check", columnDefinition = "TEXT")
    private String effectivenessCheck;

    @Column(name = "closure_summary", columnDefinition = "TEXT")
    private String closureSummary;

    @Column(name = "target_completion_date")
    private LocalDate targetCompletionDate;

    @Column(name = "submitted_by", length = 100)
    private String submittedBy;

    @Column(name = "submitted_at")
    private LocalDateTime submittedAt;

    @Column(name = "approved_by", length = 100)
    private String approvedBy;

    @Column(name = "approved_at")
    private LocalDateTime approvedAt;

    @Column(name = "approval_comments", columnDefinition = "TEXT")
    private String approvalComments;

    @Column(name = "approval_esignature_id")
    private UUID approvalESignatureId;

    @Column(name = "rejected_by", length = 100)
    private String rejectedBy;

    @Column(name = "rejected_at")
    private LocalDateTime rejectedAt;

    @Column(name = "rejection_reason", columnDefinition = "TEXT")
    private String rejectionReason;

    @Column(name = "closed_by", length = 100)
    private String closedBy;

    @Column(name = "closed_at")
    private LocalDateTime closedAt;

    @Column(name = "closure_esignature_id")
    private UUID closureESignatureId;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @Column(name = "created_by", nullable = false, length = 100)
    private String createdBy;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: core/ui/src/types/change-control.ts
import type { PageResponse } from "./grn";

export type ChangeControlStatus =
  | "DRAFT"
  | "UNDER_REVIEW"
  | "APPROVED"
  | "REJECTED"
  | "IN_IMPLEMENTATION"
  | "EFFECTIVENESS_CHECK"
  | "CLOSED"
  | "CANCELLED";

export type ChangeControlType =
  | "DOCUMENT" | "PROCESS" | "EQUIPMENT" | "MATERIAL" | "SUPPLIER" | "SYSTEM" | "OTHER";

export type ChangeControlRisk = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";

export type AffectedEntityType =
  | "MATERIAL" | "SPEC" | "MOA" | "VENDOR" | "WAREHOUSE" | "DOCUMENT" | "PROCESS" | "SYSTEM_CONFIG" | "OTHER";

export type ChangeControlTaskStatus = "PENDING" | "IN_PROGRESS" | "COMPLETED" | "SKIPPED";

export type ChangeControlAffectedEntity = {
  id: string;
  entityType: AffectedEntityType;
  entityReference: string;
  entityId: string | null;
  notes: string | null;
  createdAt: string;
};

export type ChangeControlTask = {
  id: string;
  title: string;
  description: string | null;
  assignedTo: string | null;
  dueDate: string | null;
  status: ChangeControlTaskStatus;
  completedAt: string | null;
  completedBy: string | null;
  createdAt: string;
};

export type ChangeControl = {
  id: string;
  changeControlNumber: string;
  title: string;
  description: string | null;
  changeType: ChangeControlType;
  reason: string;
  riskClassification: ChangeControlRisk;
  status: ChangeControlStatus;
  impactAssessment: string | null;
  implementationPlan: string | null;
  effectivenessCheck: string | null;
  closureSummary: string | null;
  targetCompletionDate: string | null;
  submittedBy: string | null;
  submittedAt: string | null;
  approvedBy: string | null;
  approvedAt: string | null;
  approvalComments: string | null;
  approvalESignatureId: string | null;
  rejectedBy: string | null;
  rejectedAt: string | null;
  rejectionReason: string | null;
  closedBy: string | null;
  closedAt: string | null;
  closureESignatureId: string | null;
  createdBy: string;
  createdAt: string;
  updatedBy: string | null;
  updatedAt: string | null;
  affectedEntities: ChangeControlAffectedEntity[];
  tasks: ChangeControlTask[];
};

export type CreateChangeControlRequest = {
  title: string;
  description?: string;
  changeType: ChangeControlType;
  reason: string;
  riskClassification: ChangeControlRisk;
  impactAssessment?: string;
  implementationPlan?: string;
  effectivenessCheck?: string;
  targetCompletionDate?: string;
};

export type ChangeControlApproveRequest = {
  username: string;
  password: string;
  meaning?: string;
  comments?: string;
};

export type ChangeControlCloseRequest = {
  username: string;
  password: string;
  meaning?: string;
  closureSummary: string;
};

export type AddAffectedEntityRequest = {
  entityType: AffectedEntityType;
  entityReference: string;
  entityId?: string;
  notes?: string;
};

export type CreateTaskRequest = {
  title: string;
  description?: string;
  assignedTo?: string;
  dueDate?: string;
};

export type ChangeControlPage = PageResponse<ChangeControl>;
